pkg/controller/machine-set-boot-image/vsphere_helpers.go

```diff
@@ -0,0 +1,451 @@
+		return true
+	}
+	return false
+}
+
+func findAvailableHostSystems(ctx context.Context, clusterHostSystems []*object.HostSystem, networkObjectRef object.NetworkReference, datastore *object.Datastore) (*object.HostSystem, error) {
+	var hostSystemManagedObject mo.HostSystem
+	for _, hostObj := range clusterHostSystems {
+		err := hostObj.Properties(ctx, hostObj.Reference(), []string{"config.product", "network", "datastore", "runtime"}, &hostSystemManagedObject)
+		if err != nil {
+			return nil, err
+		}
+
+		// if distributed port group the cast will fail
+		networkFound := isNetworkAvailable(networkObjectRef, hostSystemManagedObject.Network)
+		datastoreFound := isDatastoreAvailable(datastore, hostSystemManagedObject.Datastore)
+
+		// if the network or datastore is not found or the ESXi host is in maintenance mode continue the loop
+		if !networkFound || !datastoreFound || hostSystemManagedObject.Runtime.InMaintenanceMode {
+			continue
+		}
+
+		klog.V(2).Infof("using ESXi %s to import the OVA image", hostObj.Name())
+		return hostObj, nil
+	}
+	return nil, errors.New("all hosts unavailable")
+}
+
+func attachTag(ctx context.Context, tagManager *tags.Manager, vmMoRefValue, tagID string) error {
+	moRef := types.ManagedObjectReference{
+		Value: vmMoRefValue,
+		Type:  "VirtualMachine",
+	}
+
+	err := tagManager.AttachTag(ctx, tagID, moRef)
+
+	if err != nil {
+		return fmt.Errorf("unable to attach tag: %w", err)
+	}
+	return nil
+}
+
+// Used govc/importx/ovf.go as an example to implement
+// resourceVspherePrivateImportOvaCreate and upload functions
+// See: https://github.com/vmware/govmomi/blob/cc10a0758d5b4d4873388bcea417251d1ad03e42/govc/importx/ovf.go#L196-L324
+func upload(ctx context.Context, archive *importer.TapeArchive, lease *nfc.Lease, item nfc.FileItem) error {
+	file := item.Path
+
+	f, size, err := archive.Open(file)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	opts := soap.Upload{
+		ContentLength: size,
+	}
+
+	return lease.Upload(ctx, item, f, opts)
+}
+
+func findAllRequiredResources(ctx context.Context, finder *find.Finder, providerSpec *machinev1beta1.VSphereMachineProviderSpec, failureDomain osconfigv1.VSpherePlatformFailureDomainSpec) (*object.Folder, *object.ClusterComputeResource, *object.ResourcePool, object.NetworkReference, *object.Datastore, error) {
+	folder, err := finder.Folder(ctx, providerSpec.Workspace.Folder)
+	if err != nil {
+		return nil, nil, nil, nil, nil, fmt.Errorf("failed to find folder: %w", err)
+	}
+	cluster, err := finder.ClusterComputeResource(ctx, failureDomain.Topology.ComputeCluster)
+	if err != nil {
+		return nil, nil, nil, nil, nil, fmt.Errorf("failed to find compute cluster: %w", err)
+	}
+	resourcePool, err := finder.ResourcePool(ctx, providerSpec.Workspace.ResourcePool)
+	if err != nil {
+		return nil, nil, nil, nil, nil, fmt.Errorf("failed to find resource pool: %w", err)
+	}
+	networkPath := path.Join(cluster.InventoryPath, failureDomain.Topology.Networks[0])
```

Comment:
be aware in 4.20 multiple nics will be ga

Reply:
I have mirrored the implementation over https://github.com/openshift/installer/blob/0b51991c838f746e473a6f8a0982716028d76405/pkg/infrastructure/vsphere/clusterapi/import.go#L115
Will update as per your guidance when multiple NICs GA